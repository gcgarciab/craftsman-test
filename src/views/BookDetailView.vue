<template>
  <section class="book-detail wrapper">
    <div v-if="!isSelectedBookEmpty" class="book-info items-container">
      <AppButton
        icon="heart"
        class="absolute top-5 right-5"
        title="Go to favorites"
        :type="ButtonType.BUTTON"
        :color="ButtonColor.DANGER"
        @click="$router.push({ name: 'Favorites' })"
      />

      <h2 class="title">Book detail</h2>

      <div class="volume-info item">
        <BookCard
          :id="currentBook.id"
          :volume-info="currentBook.volumeInfo"
          :is-favorite="favorites.includes(currentBook.id)"
          @toggle-favorite="toggleFavoriteBook(currentBook.id)"
        />
      </div>

      <div v-if="volumeInfo.description" class="description item">
        <h4 class="subtitle">Description</h4>
        <div
          class="description-data text-left px-5"
          v-html="volumeInfo.description"
        />
      </div>

      <div v-if="volumeInfo.categories?.length" class="categories item">
        <h4 class="subtitle">Categories</h4>

        <p
          v-for="(category, index) in volumeInfo.categories"
          :key="`category-${index + 1}`"
          class="category px-5"
        >
          {{ category }}
        </p>
      </div>

      <div class="book-links item">
        <h4 class="subtitle">Quick links</h4>

        <div class="links">
          <div class="book-link">
            <AppButton
              text="Preview"
              icon="book"
              test-name="preview-link"
              @click="goToLink(volumeInfo.previewLink!)"
            />
          </div>

          <div class="book-link">
            <AppButton
              text="Info"
              icon="book"
              test-name="info-link"
              @click="goToLink(volumeInfo.infoLink!)"
            />
          </div>

          <div class="book-link">
            <AppButton
              text="Canonical"
              icon="book"
              test-name="canonical-link"
              @click="goToLink(volumeInfo.canonicalVolumeLink!)"
            />
          </div>
        </div>
      </div>

      <div
        v-if="saleInfo.saleability !== BookSaleability.NOT_FOR_SALE"
        class="sale-info item"
      >
        <h4 class="subtitle">Price</h4>

        <div class="prices px-2 lg:px-5">
          <div class="price-items">
            <p v-if="saleInfo.listPrice?.amount" class="book-price">
              <span>List Price:</span>
              <BookPrice :price="saleInfo.listPrice" />
            </p>

            <p v-if="saleInfo.retailPrice?.amount" class="book-price">
              <span>Retail Price:</span>
              <BookPrice :price="saleInfo.retailPrice" />
            </p>

            <div v-if="saleInfo.buyLink" class="sale-action">
              <AppButton
                icon="money-bill-wave"
                text="Buy"
                test-name="buy-link"
                :type="ButtonType.BUTTON"
                :color="ButtonColor.SUCCESS"
                @click="goToLink(saleInfo.buyLink!)"
              />
            </div>
          </div>
        </div>
      </div>

      <BackToHomeButton />
    </div>

    <template v-else>
      <AppLoader v-if="mainStore.isLoading" />

      <ErrorMessage
        v-else
        title="Error!"
        description="Not found data from current book"
        icon="exclamation-triangle"
        :back-to-home="true"
      />
    </template>
  </section>
</template>

<script setup lang="ts">
import { storeToRefs } from 'pinia';
import { useRoute } from 'vue-router';
import { computed, onBeforeMount, reactive } from 'vue';

import useBookStore from '@/stores/book';
import useMainStore from '@/stores/main';
import Book from '@/interfaces/book';
import ButtonType from '@/enums/button-type';
import ButtonColor from '@/enums/button-color';
import BookSaleability from '@/enums/book-saleability';
import AppLoader from '@/ui/AppLoader.vue';
import AppButton from '@/ui/AppButton.vue';
import BookCard from '@/components/BookCard.vue';
import BookPrice from '@/components/BookPrice.vue';
import ErrorMessage from '@/components/ErrorMessage.vue';
import BackToHomeButton from '@/components/BackToHomeButton.vue';

const route = useRoute();
const bookStore = useBookStore();
const mainStore = useMainStore();
const currentBook = reactive<Book>({} as Book);
const { favorites } = storeToRefs(bookStore);
const bookId = route.params.id as string;

const isSelectedBookEmpty = computed(
  () => Object.keys(currentBook).length === 0
);

const saleInfo = computed(() => currentBook.saleInfo);
const volumeInfo = computed(() => currentBook.volumeInfo);

/**
 * Open Book URL in a new tab.
 * @param url => Buy URL
 */
function goToLink(url: string): void {
  console.log('Click !!');
  window.open(url, '_blank');
}

/**
 * Calls toggleFavoriteBook action from store to add or remove
 * specific book from favorites state.
 * @param id => Book id
 */
function toggleFavoriteBook(id: string) {
  bookStore.toggleFavoriteBook(id);
}

onBeforeMount(async () => {
  mainStore.$patch({ isLoading: true });

  const data = await bookStore
    .fetchBookById(bookId)
    .catch(() => mainStore.$reset());

  Object.assign(currentBook, data);
  mainStore.$reset();
});
</script>

<style scoped>
.item {
  @apply my-10 text-left w-full;
}

.description-data :deep(p) {
  @apply my-6;
}

.price-items {
  @apply flex flex-col justify-between border md:flex-row;
}

.price-items > :nth-child(2) {
  @apply border-y;
  @apply md:border-x md:border-y-0;
}

.price-items .book-price {
  @apply w-full py-6 flex justify-center items-center;
  @apply md:w-1/3;
}

.sale-action {
  @apply w-full px-10 m-auto flex flex-col justify-center my-5;
  @apply sm:w-3/4;
  @apply md:w-1/3 md:flex-row;
}

.links {
  @apply flex w-full flex-col items-center;
  @apply sm:flex-wrap sm:flex-row;
}

.book-link {
  @apply w-full flex flex-col items-stretch px-16 my-4;
  @apply sm:px-2 sm:w-1/3;
  @apply md:px-4 md:my-0;
  @apply lg:px-10;
  @apply xl:px-16;
}
</style>
