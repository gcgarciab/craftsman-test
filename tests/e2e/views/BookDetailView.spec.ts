import { mount } from '@vue/test-utils';
import { createTestingPinia, TestingOptions } from '@pinia/testing';
import flushPromises from 'flush-promises';
import useBookStore from '@/stores/book';
import findById from '@/utils/find-by-id';
import favoritesData from '@/mocks/books-by-id.json';
import BookDetailView from '@/views/BookDetailView.vue';

const fakeSelectedId = 'VvJZ3lp2sg0C';

jest.mock('vue-router', () => ({
  useRoute: jest.fn(() => ({ name: 'BookDetail', params: { id: 'VvJZ3lp2sg0C' } })),
  useRouter: jest.fn(() => ({ push: jest.fn() })),
}));

jest.mock('axios', () => ({
  create: jest.fn(() => ({
    get: jest.fn(() => Promise.resolve({ data: findById(favoritesData, fakeSelectedId) })),
    interceptors: {
      request: { use: jest.fn(), eject: jest.fn() },
      response: { use: jest.fn(), eject: jest.fn() },
    },
  })),
}));

describe('BookDetailView', () => {
  function factory(options?: TestingOptions) {
    const wrapper = mount(BookDetailView, {
      global: {
        plugins: [
          createTestingPinia(options),
        ],
        mocks: { $route: { 
          name: 'BookDetail',
          params: { id: fakeSelectedId } } }
      },
    });

    const store = useBookStore();

    return { wrapper, store };
  }

  it('Read fakeSelectedId bookId great', async () => {
    const { store, wrapper } = factory({ stubActions: false });

    expect(wrapper.vm.$route.name).toBe('BookDetail');
    expect(wrapper.vm.$route.params.id).toBe('VvJZ3lp2sg0C');
  });

  it('Call fetchBookById action from store', async () => {
    const { store, wrapper } = factory({ stubActions: false });

    await flushPromises();

    expect(store.fetchBookById).toBeCalled();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Call toggleFavoriteBook action from store', async () => {
    const { store, wrapper } = factory({ stubActions: false });

    await flushPromises();

    const favoriteButton = wrapper.find('[data-test="add-to-favorites"]');
    await favoriteButton.trigger('click');

    expect(store.favorites.length).toBe(1);

    await favoriteButton.trigger('click');
    expect(store.favorites.length).toBe(0);
  });

  it('Call goToLink method on detail and buy buttons', async () => {
    const { store, wrapper } = factory({ stubActions: false });
    const spy = jest.spyOn(window, 'open');

    await flushPromises();

    const previewButton = wrapper.find('[data-test="preview-link"]');
    await previewButton.trigger('click');
    // expect(spy).toBeCalledTimes(1);
    
    // const buyButton = wrapper.find('[data-test="buy-link"]');
    // await buyButton.trigger('click');
    // expect(spy).toBeCalledTimes(2);
  });
})