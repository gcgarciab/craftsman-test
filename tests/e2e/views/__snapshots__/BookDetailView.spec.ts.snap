// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookDetailView Call fetchBookById action from store 1`] = `
"<section class=\\"book-detail wrapper\\">
  <div class=\\"book-info items-container\\"><button class=\\"text-white px-2 flex items-center justify-center danger absolute top-5 right-5\\" data-test=\\"button\\" type=\\"button\\" title=\\"Go to favorites\\">
      <fa class=\\"icon text-sm\\" icon=\\"heart\\"></fa>
    </button>
    <h2 class=\\"title\\">Book detail</h2>
    <div class=\\"volume-info item\\">
      <div class=\\"book-card\\">
        <div class=\\"book-header\\">
          <div class=\\"book-image w-36 h-48 flex justify-center items-center border\\"><img src=\\"http://books.google.com/books/content?id=VvJZ3lp2sg0C&amp;printsec=frontcover&amp;img=1&amp;zoom=5&amp;imgtk=AFLRE71xS6JObSmwNwrTfoDr8Sz1zoFuOllqscsfjDaWHrRcI3iznK8QFhq4-TD7gd3m7MmgEoyhdIYY5UXcHRY7z-ZzM1nmNZ-b3cyBqbCUXgeXslF3DfnAMUSD5W6SgIHfbDb8bpVf&amp;source=gbs_api\\" alt=\\"Book image\\" title=\\"Book image\\"></div>
          <div class=\\"main-info\\">
            <h3 class=\\"book-title\\">SEAFOOD COOK BOOK
              <!--v-if-->
            </h3>
            <h4 class=\\"authors\\"></h4>
            <!--v-if-->
            <div class=\\"language mt-3\\"> Language: (EN) <span class=\\"total-pages mt-3\\"> - Pages: 112</span></div>
          </div>
        </div>
        <div class=\\"card-actions absolute right-6 top-6 w-20 flex justify-end\\">
          <!--v-if--><button class=\\"text-white px-2 flex items-center justify-center default\\" data-test=\\"add-to-favorites\\" type=\\"button\\" title=\\"Add to favorites\\">
            <fa class=\\"icon text-sm\\" icon=\\"heart\\"></fa>
          </button>
        </div>
      </div>
    </div>
    <!--v-if-->
    <!--v-if-->
    <div class=\\"book-links item\\">
      <h4 class=\\"subtitle\\">Quick links</h4>
      <div class=\\"links\\">
        <div class=\\"book-link\\"><button class=\\"text-white px-10 flex items-center justify-center default\\" data-test=\\"preview-link\\" type=\\"button\\">
            <fa class=\\"icon text-sm mr-2\\" icon=\\"book\\"></fa> Preview
          </button></div>
        <div class=\\"book-link\\"><button class=\\"text-white px-10 flex items-center justify-center default\\" data-test=\\"info-link\\" type=\\"button\\">
            <fa class=\\"icon text-sm mr-2\\" icon=\\"book\\"></fa> Info
          </button></div>
        <div class=\\"book-link\\"><button class=\\"text-white px-10 flex items-center justify-center default\\" data-test=\\"canonical-link\\" type=\\"button\\">
            <fa class=\\"icon text-sm mr-2\\" icon=\\"book\\"></fa> Canonical
          </button></div>
      </div>
    </div>
    <!--v-if--><button class=\\"text-white mt-8 px-10 flex items-center justify-center primary\\" data-test=\\"button\\" type=\\"button\\">
      <fa class=\\"icon text-sm mr-2\\" icon=\\"home\\"></fa> Back to home
    </button>
  </div>
</section>"
`;
